### PR TITLE
chore(ion): add description in readme for outline

### PR DESCRIPTION
**Pull request by BenjaminMoe**
Creating a draft pull request and then using that as a notepad to work my way through a problem seems to be the best approach I've found so far. In this case we need to get a local version of Ion running so that we can get test vectors from it. 

Let's take a look at the easy part. To install Ion, we need the dependencies. In general this just means having 
- Bitcoin-core
- IPFS
- MongoDB
- Nodejs 14

Installed on our Pi. The hard part is going to be how to use these. Specifically we have two tasks to work with. 

The first task we have is to get Ion working with Regtest. This isn't too hard on the Ion side of things as it looks like there is already a config for working with Regtest in the example configs. What's going to be challenging is addressing how to work with Regtest. Will we need to set up multiple nodes? Will we need to mine? And how do we manage our wallet?

The second task is how to actually use Ion, in terms submitting a create operation and then logging the hashes and the core-index file. So we might need to read through the code and the CLI to try and figure out what needs to be called and where to make a request.

In terms of complexity it might be easier to work with the testnet as there is element of management that we don't need to deal with. We get a small amount of testcoins from a faucet and then we write to the network and log as needed. But in terms of testing and reproducibility we should probably be working with Regtest so that we can start from a specific point, and make tests as needed.

### CODE REVIEW

**Comment by BenjaminMoe** [comment]
First notes in the stream of consciousness. We got the dependencies installed pretty quickly because we were able to copy and paste from our previous document. 

Before we do anything with Ion, the next step will be to figure out how to use Regtest. From what I can find of documents that actually describe how to use it, it looks pretty simple. It looks like we use a command to generate blocks. And in the same command we can decide where the mined coins from the block are deposited to.

From what I can guess is that is we don't define an address, it will be put into the address of out bitcoin-cli account. I guess we'll see. I'm not exactly sure where any Bitcoin is sent when we publish a new anchor string to the block chain, but I guess we'll find about that as well. The main thing that we should focus on next is playing around with Regtest so that we can get familiar with it. We can always delete out `~/.bitcoin` folder to start over if we need to.

**Comment by BenjaminMoe** [comment]
Time to try and cut through the junk with all of the `HoW to MiNE BiTCoiN oN RAsPbErRy Pi` search results, and take notes on how to run Regtest on a Raspberry Pi, or just in general. 

Normally I would do this in screen, but in general we want to be able to scroll easily, so I'll just do it in a normal console. We run 
```$ bitcoin-core -regtest```
We shouldn't need to do too much else. In this case any parameters for `bitcoin-core` can be defined as command line arguments. We'll avoid trying to be cool kids and trying to write a config file that does things for us. We can make that transparent by showing which arguments are needed, when, and potentially why.

**Comment by BenjaminMoe** [comment]
I'm not too worried about sending transactions on the command line, or generating multiple wallets. As long as we have an address and we have some amount of bitcoin that's spendable, we can go ahead and progress. The next step we need to focus on is running the bitcoin process for Ion. This means we need to set up the config file, and then run the process. 

So first, let's break down the config file to see if that offers any hints, and then let's see what starting the process will do.

**Comment by BenjaminMoe** [comment]
Now we're finally to the point where we can run the bitcoin-ion process. First we need to build it with 

```
$ cd ~/ion
$ npm i
$ npm run build
```

Will create the "binary" which can be found in `dist/bin`. Not sure what the command is to run a development process directly from the typescript source files, but for now we'll go with what's in the `package.json` file and be grateful that it even runs in the first place. Since we need to change the config files, we'll go ahead and do that inside of the `dist` directory.

```
$ rm -rf dist/json/*
$ vim dist/json/testnet-core-config.json
--- Create File ---
{
  "bitcoinDataDirectory": "/home/ubuntu/snap/bitcoin-core/common/.bitcoin/regtest",
  "bitcoinFeeSpendingCutoffPeriodInBlocks": 1,
  "bitcoinFeeSpendingCutoff": 0.001,
  "bitcoinPeerUri": "http://localhost:18443",
  "bitcoinRpcUsername": "",
  "bitcoinRpcPassword": "",
  "bitcoinWalletOrImportString": "bcrt1qjglqvdy353kwrrgx8kljva26v60ltuwwtgw4eh",
  "databaseName": "ion-regtest-bitcoin",
  "genesisBlockNumber": 100,
  "logRequestError": true,
  "mongoDbConnectionString": "mongodb://localhost:27017/",
  "port": 3002,
  "sidetreeTransactionFeeMarkupPercentage": 1,
  "sidetreeTransactionPrefix": "ion:",
  "transactionPollPeriodInSeconds": 60,
  "valueTimeLockUpdateEnabled": false,
  "valueTimeLockAmountInBitcoins": 0,
  "valueTimeLockPollPeriodInSeconds": 600 ,
  "valueTimeLockTransactionFeesAmountInBitcoins": 0.0001
}
--- EOF ---
$ vim dist/json/regtest-bitcoin-versioning.json
--- Create File ---
[
  {
    "startingBlockchainTime": 100,
    "version": "latest",
    "protocolParameters": {
      "feeLookBackWindowInBlocks": 1000,
      "feeMaxFluctuationMultiplierPerBlock": 0.00001,
      "initialNormalizedFeeInSatoshis": 400,
      "valueTimeLockDurationInBlocks": 150
    }
  }
]
--- EOF ---
```

I guess the `json` folder from the base directory is being copied into `dist`. So it looks like it would be a better idea to define the files in the root directory so they can be copied over. In the current context, we only need to run ion, and not specifically edit anything. So it should be okay. I don't don't see myself changing the config too often. Only small issue that I'm kind of annoyed about is not having a generic path for the default config file. 

In other words, there is regtest, testet, and main-net. So I'm kind of surprised that the config defaults to testnet. Well not entirely surprised, seems like something the devs did to save themselves on time because that's what they were using the most. But in terms of convention, normally you would expect to have options, and then a default generic config name for the one that gets used. The config file naming is soon going to be the least of my concern, so I'll go ahead and copy it over and move on to the next thing, assuming that works.

**Comment by BenjaminMoe** [comment]
Looks like i was incorrect in thinking that the `bitcoinWalletOrImportString` attribute was the wallet address. Which makes sense now that I think about it. What's needed for the node is to create transactions, which probably means it needs the private key for the address. Which means this is probably the seed string for the wallet. 

We can get this information with
```
$ bitcoin-core.cli -regtest getwalletinfo
{
  "walletname": "ion-dev",
  "walletversion": 169900,
  "format": "bdb",
  "balance": 300.00000000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 5000.00000000,
  "txcount": 106,
  "keypoololdest": 1637340019,
  "keypoolsize": 999,
  "hdseedid": "79c51bd6148b4c446c4887c0c047315312c190d7",
  "keypoolsize_hd_internal": 1000,
  "paytxfee": 0.00000000,
  "private_keys_enabled": true,
  "avoid_reuse": false,
  "scanning": false,
  "descriptors": false
}
```

So maybe `hdseedid`?

```
$ vim ../json/testnet-bitcoin-config.json
--- Create File ---
{
  "bitcoinDataDirectory": "/home/ubuntu/snap/bitcoin-core/common/.bitcoin/regtest",
  "bitcoinFeeSpendingCutoffPeriodInBlocks": 1,
  "bitcoinFeeSpendingCutoff": 0.001,
  "bitcoinPeerUri": "http://localhost:18443",
  "bitcoinRpcUsername": "",
  "bitcoinRpcPassword": "",
  "bitcoinWalletOrImportString": "79c51bd6148b4c446c4887c0c047315312c190d7",
  "databaseName": "ion-regtest-bitcoin",
  "genesisBlockNumber": 100,
  "logRequestError": true,
  "mongoDbConnectionString": "mongodb://localhost:27017/",
  "port": 3002,
  "sidetreeTransactionFeeMarkupPercentage": 1,
  "sidetreeTransactionPrefix": "ion:",
  "transactionPollPeriodInSeconds": 60,
  "valueTimeLockUpdateEnabled": false,
  "valueTimeLockAmountInBitcoins": 0,
  "valueTimeLockPollPeriodInSeconds": 600 ,
  "valueTimeLockTransactionFeesAmountInBitcoins": 0.0001
}
--- EOF ---
```

And we get something different

```
Creating bitcoin wallet using the import string passed in.
Sidetree bitcoin service configuration:
(node:21588) Warning: Accessing non-existent property 'MongoError' of module exports inside circular dependency
(Use `node --trace-warnings ...` to show where the warning was created)
(node:21588) DeprecationWarning: current Server Discovery and Monitoring engine is deprecated, and will be removed in a future version. To use the new Server Discover and Monitoring engine, pass option { useUnifiedTopology: true } to the MongoClient constructor.
Collection 'service' does not exists, creating...
Collection 'service' created.
Collection 'blocks' does not exists, creating...
Collection 'blocks' created.
Transaction collection does not exists, creating...
Transaction collection created.
Sending jRPC request: id: 3heojb98flc, method: createwallet
Sidetree-Bitcoin node initialization failed with error: {"stack":"Error: Fetch failed [401]: \n    at BitcoinClient.<anonymous> (/home/ubuntu/ion/node_modules/@decentralized-identity/sidetree/dist/lib/bitcoin/BitcoinClient.js:755:31)\n    at Generator.next (<anonymous>)\n    at fulfilled (/home/ubuntu/ion/node_modules/@decentralized-identity/sidetree/dist/lib/bitcoin/BitcoinClient.js:5:58)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)","message":"Fetch failed [401]: "}
```

So two options. First is that it fails on the first time and we try again. Second option is we probably need a server=1 option on the bitcoint-core daemon process.

**Comment by BenjaminMoe** [comment]
It looks like using my existing wallet does work because automatically tries to create a wallet?

```
Sending jRPC request: id: 5n7na65senv, method: createwallet
http://localhost:18443
{
  body: '{"method":"createwallet","params":["sidetreeDefaultWallet"],"jsonrpc":"1.0","id":"5n7na65senv"}',
  method: 'post'
}
Sidetree-Bitcoin node initialization failed with error: {"stack":"Error: Fetch failed [401]: \n    at BitcoinClient.<anonymous> (/home/ubuntu/ion/node_modules/@decentralized-identity/sidetree/dist/lib/bitcoin/BitcoinClient.js:759:31)\n    at Generator.next (<anonymous>)\n    at fulfilled (/home/ubuntu/ion/node_modules/@decentralized-identity/sidetree/dist/lib/bitcoin/BitcoinClient.js:5:58)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)","message":"Fetch failed [401]: "}
```

Which means I guess I might need to use the automatically generated string? And then send coins to that address?

**Comment by BenjaminMoe** [comment]
I think we've managed to get a minimum viable configuration to get the bitcoin-ion process running. I think the current context we will write the config file and set up ion-bitcoin, and ion-core running in their own respective screen processes, so that we can move onto writing and then reading dids from the ion service. 

In terms of documentation, we should go ahead and make notes, even if they are incomplete so we don't forget anything while we're thinking about it. We can come but and update the details when we comeback for another confirmation pass. The good thing about working with regtest is that we don't need to compile or sync a lot of data from the network, so starting fresh for confirmation is not that that time-consuming of an activity.

**Comment by BenjaminMoe** [comment]
I think we managed to get the ion-bitcoin and ion-core processes running with regtest. And we have a wallet with a private key defined in the configs. In retrospect there wasn't a lot too this. I think it was mostly the requirement was familiarity with bitcoin-cli. And I think the devs assumed that if anyone would be using Regtest, it was probably be someone already familiar with bitcoin. I don't fit into that category, but I understand the reasoning behind that assumption. 

What's next is being able to create, and then resolve a did using the local testnet. The next concern I had was the lack of documentation for any documentation on specifically how to interact with the API. But now that I think about it, since they're implementing the sidetree protocol, we should have specific end points that we can work with. That means we can probably used a fixed example for testing. If I get stuck I can sync with Orie on what parameters the API expects, but I should give it a try on my own before that.

**Comment by BenjaminMoe** [comment]
Moving on to the next step of how we can send a create operation to the API. With respect to this, there are two main concerns I have. The first being the format needed to send a request to the node in the first place. And the second being the timing needed to create a new block on the Regtest blockchain.

With respect to the second concern, I think we have two options. One is that we could write a short script that would send an http reqest to the server and then trigger a generate block command. Or otherwise we could potentially have another screen where we simply create a new block every 30 seconds, or otherwise some period of time that Ion expects a new block.

The lack of documentation with respect to the CRUD operations is the most confusing in my option. I think I'll have to check out the sidetree spec and see what the expected format is there. But while I'm ranting, I'll point out two more points of confusion. One being that the Ion directory contains an `ion` folder that contains http files. Maybe it doesn't have any special functionality so it's not worth mentioning in the readme files, but it's still weird that it's not mentioned. And the second point of curiosity is that Ion has a command line utility if you run `npm install -g`, which again is not mentioned in the documentation as far as I can tell. 

Ranting aside, let's check out the source code for the ion-core process and the specification on sidetree to see if that helps us.

**Comment by BenjaminMoe** [comment]
So first step, popping into `src/core.ts` we can take a look at the ion-core API, which should be listening on port 3000. We see exactly what we would expect to see here. 

```
const router = new Router();
router.post('/operations', async (ctx, _next) => {
  const response = await sidetreeCore.handleOperationRequest(ctx.body);
  setKoaResponse(response, ctx.response);
});

router.get('/version', async (ctx, _next) => {
  const response = await sidetreeCore.handleGetVersionRequest();
  setKoaResponse(response, ctx.response);
});

const resolvePath = '/identifiers/';
router.get(`${resolvePath}:did`, async (ctx, _next) => {
  // Strip away the first '/identifiers/' string.
  const didOrDidDocument = ctx.url.split(resolvePath)[1];
  const response = await sidetreeCore.handleResolveRequest(didOrDidDocument);
  setKoaResponse(response, ctx.response);
});
```

We have `/operations` and `/identifiers` which are defined in the sidetree protocol spec. So let's try sending a curl request and see if we denied, and what comes back. 

```
$ curl -X POST localhost:3000/operations
```

And then we can log in the application to make sure we get a response, which we did.

```
OPERATIONS WAS CALLED!!!!
Handling operation request of size 0 bytes...
Bad request: [object Object]
```

So the next question is "what is a valid request?"

**Comment by BenjaminMoe** [comment]
The API definition for sidetree can be found here: https://identity.foundation/sidetree/api/#create.

```
{
    "type": "create",
    "suffixData": SUFFIX_DATA_OBJECT,
    "delta": DELTA_OBJECT
}
```

This is the format that the post data needs to be in. We need to figure out how to make a simple-as-possible "hello world" object. Which gives us two questions. How do we compute the suffix data, and what's a minimum viable delta object?

**Comment by BenjaminMoe** [comment]
Let's focus on a simple proof of concept. We'll manually send a single transaction, and then generate multiple blocks with regtest via the command line. For a create operation we have the following command prepared.

```
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"type":"create","suffixData":{"deltaHash":"EiDqw2eLXZqH8PkFgdWuse-3RrkLZtFhiZElPBO2Vizm1A","recoveryCommitment":"EiCwadKJRWbjRXZ0Ug_jTlM2z0eV0lpVuGrpv4q9t0W0uw"},"delta":{"updateCommitment":"EiAR_MRq4eKPumuLJyLmxRUscIfVjolBgOVhUfl9yl1MGA","patches":[{"action":"replace","document":{"publicKeys":[{"id":"signing-key","type":"EcdsaSecp256k1VerificationKey2019","publicKeyJwk":{"kty":"EC","crv":"secp256k1","x":"NnIQIF0j9bPnsRxxf7Imae9YbK9KJFeU_YZfoA3UwKo","y":"GEzrxwz2qf6w04s31p-ohPgkkb8MGE1dO7RtVIAgxKg"}}]}}]}}' \
  http://localhost:3000/operations
```

To create blocks we can run 
```
bitcoin-core.cli -generate 1
```

Let's see how this goes and if we can resolve our did.

**Comment by BenjaminMoe** [comment]
Digging into the error a little bit, let's try and use the bitcoin-cli to debug what's going on with our wallet. 

```
ubuntu@ubuntu:~$ bitcoin-core.cli listreceivedbyaddress 1 true
[
  {
    "address": "bcrt1qz9x6kgvf4y7ell5t322aa7w3qss6rgksnnrlw5",
    "amount": 0.00000000,
    "confirmations": 0,
    "label": "",
    "txids": [
    ]
  },
  {
    "address": "bcrt1qx7jm96nh280uq2ruhjakm395x4gr7ryvcudu9a",
    "amount": 0.00000000,
    "confirmations": 0,
    "label": "",
    "txids": [
    ]
  },
  {
    "address": "bcrt1qdl2q58y4ll5qvyle3jj2vp7ju76redqqjzclrk",
    "amount": 0.00000000,
    "confirmations": 0,
    "label": "",
    "txids": [
    ]
  },
  {
    "address": "bcrt1qjglqvdy353kwrrgx8kljva26v60ltuwwtgw4eh",
    "amount": 0.00000000,
    "confirmations": 0,
    "label": "",
    "txids": [
    ]
  }
]
```

So it says that there's no coins in the wallet? That would explain why Ion is throwing an error. but that's kind of weird if the `getbalance` shows that we have coins, but then the list of addresses doesn't have any coins? I guess we need to do some more trial and error to see about getting coins into a specific address so that we can dump the private key to put into the ion config.

**Comment by BenjaminMoe** [comment]
This time I tried specifically sending funds to the wallet address from the wallet. Ion seems to have gotten the `create` operation, and hasn't thrown an error.... yet. 

```
OPERATIONS WAS CALLED!!!!
Handling operation request of size 512 bytes...
Operation type: 'create', DID unique suffix: 'EiBpAx8f44rlz84PrnqecY0FNBUM-ufEUWS4Pqi36uYu_g'
Getting blockchain time...
Got latest blockchain time: {"time":516,"hash":"6db573fbbf59e43c8515bf01b10d460ecdb8fba10dad71651a540eb9ae4ca6b4"}
Core cachedApproximateTime updated to: 516
Fetching Sidetree transactions from blockchain service...
Fetching URI 'http://127.0.0.1:3002/transactions'...
Fetch response: 200'.
Fetched 0 Sidetree transactions from blockchain service in 37 ms.
Successfully kicked off downloading/processing of all new Sidetree transactions.
Processing previously unresolvable transactions if any...
Fetched 0 unresolvable transactions to retry in 3 ms.
Event emitted: sidetree_observer_loop_success
Waiting for 60 seconds before fetching and processing transactions again.
```

**Comment by BenjaminMoe** [comment]
Taking a quick step back. There's a possibility that the wallet name is the only thing holding me back from a basic read/write operation. But before I test that I want to make sure to start syncing the testnet on another Pi. I was thinking that it might be a good contrast for regtest, but I think the wallet name issue might popup everywhere. So I'm not as convinced as before. 

Either way, I should get bitcoin synced so I can install Ion and test that later. It seems like whenever I try to start ion, that it will try to write the pending operation. So this gives me the chance to log the CAS interface to see how they are handling the core-index file.

**Comment by OR13** [comment]
@BenjaminMoe You have leaked your local node basic auth password: `YWRtaW46a2V5Ym9hcmRjYXQ=` -> `admin:keyboardcat`... make sure that does not matter to you :)

**Comment by BenjaminMoe** [comment]
> @BenjaminMoe You have leaked your local node basic auth password: `YWRtaW46a2V5Ym9hcmRjYXQ=` -> `admin:keyboardcat`... make sure that does not matter to you :)

Yup, it's okay. This is regtest only running a completely separate chain locally on my computer, so these values don't matter. My plan is to make security practices lax for local, strict for testnet, super strict for mainnet.

**Comment by BenjaminMoe** [comment]
Random notes is getting too long. Starting a new PR.
